### ISSUE TITLE
RTSL does not save Texture2D linked to Normal Map field of material using URP

### ISSUE THREAD

**Issue by DHeroPlasma** [Battlehub0x/RTE_Docs#79]
I have a problem with RTSL and normal maps. I am using URP and the normal map is of type „Texture2D“. In my RTSL configuration, I have checked „Texture2D“ as a persistent class. Am I missing something? Because after reloading, the texture file under „normal map“ is gone in my attached material (see attached screenshot).

I am using:
- Unity 2020.3.33f1
- Battlehub/RTE/RTSL version 2.11
- Universal Render Pipeline

Also, building persistent classes for all kinds of URP classes did not help.

Thanks for any kind of help!

![URP_NormalMap_RTSL-Bug](https://github.com/Battlehub0x/RTE_Docs/assets/60179146/0f6c1970-ec5c-4e93-8ab3-c5f025bd53b5)

**Comment by Battlehub0x** [comment]
Hi @DHeroPlasma, here is the solution to this problem:
0. Upgrade to RTSL 3.5.0 (optional);
1. Install Burst and Mathematics packages:
![image](https://github.com/Battlehub0x/RTE_Docs/assets/15802443/253e098f-b180-4fdd-a986-914baece90fe)

2. Create following extension:
```
using Unity.Burst;
using Unity.Collections;
using Unity.Jobs;
using Unity.Mathematics;
using UnityEngine;

namespace Battlehub.Utils
{
    public static class IsNormalMapExtension
    {
        [BurstCompile]
        public struct CountUnitVectorsJob : IJobParallelFor
        {
            [ReadOnly] public NativeArray<Color> Colors;

            [ReadOnly] public int Offset;

            public NativeArray<int> Result;

            public void Execute(int index)
            {
                Color color = Colors[index];

                float3 n = new float3(color.a * 2 - 1, color.g * 2 - 1, 0);
                n.z = math.sqrt(1 - math.dot(n, n));

                float3 normal = new float3(n.x, n.z, n.y);
                Result[index] = math.abs(math.lengthsq(normal) - 1) < 0.00001f ? 1 : 0;
            }
        }

        [BurstCompile]
        public struct SumUnitVectorsJob : IJob
        {
            public NativeArray<int> UnitVectors;

            public NativeArray<int> Result;

            public void Execute()
            {
                int sum = 0;
                for (int i = 0; i < UnitVectors.Length; ++i)
                {
                    sum += UnitVectors[i];
                }

                Result[0] = sum;
            }
        }

        public static bool IsNormalMap(this Texture2D texture)
        {
            //var sw = System.Diagnostics.Stopwatch.StartNew();
            int w = texture.width;
            int h = texture.height;

            var pixels = texture.GetPixels();
            using var colors = new NativeArray<Color>(pixels, Allocator.TempJob);
            using var unitVectors = new NativeArray<int>(pixels.Length, Allocator.TempJob);
            using var result = new NativeArray<int>(1, Allocator.TempJob);

            var countVectorsJob = new CountUnitVectorsJob
            {
                Colors = colors,
                Result = unitVectors,
            };

            var jobHandle = countVectorsJob.Schedule(countVectorsJob.Colors.Length, 100);
            var sumJob = new SumUnitVectorsJob
            {
                UnitVectors = unitVectors,
                Result = result,
            };

            jobHandle = sumJob.Schedule(jobHandle);
            jobHandle.Complete();

            bool isNormalMap = result[0] > w * h * 0.8f;

            //Debug.Log("IsNormalMapCPUJob: " + sw.ElapsedMilliseconds + " ms");
            //Debug.Log("IsNormalMap: " + isNormalMap);

            return isNormalMap;
        }
    }
}
```
3. Modify PersistentTexture2D.cs 
![image](https://github.com/Battlehub0x/RTE_Docs/assets/15802443/4fdd7f86-215f-472e-92e3-202d9f19ab07)
![image](https://github.com/Battlehub0x/RTE_Docs/assets/15802443/4daf65f5-8dc7-4a9c-af84-626811ca8040)
![image](https://github.com/Battlehub0x/RTE_Docs/assets/15802443/7f909398-617e-4950-9735-3fe396973234)

After that, the normal map textures will be loaded and rendered correctly. Note that in the Unity Inspector you will still see the texture with a pink tint.

**Comment by DHeroPlasma** [comment]
Works like a charm

**Comment by Battlehub0x** [comment]
I forgot to mention that you need to click the "Build All" button, see the picture below.

![image](https://github.com/Battlehub0x/RTE_Docs/assets/15802443/c4692369-0d9e-4b20-98ec-88d95b96038e)

RTSL uses [protobuf.net](https://github.com/mgravell/protobuf-net) for serialization.

RTSLTypeModel.dll contains [protobuf-net](https://github.com/mgravell/protobuf-net) type model. Due to [unity scripting restrictions](https://docs.unity3d.com/Manual/ScriptingRestrictions.html), runtime type model need to be pre-complied before using at runtime.

![image](https://github.com/Battlehub0x/RTE_Docs/assets/15802443/0d44a5c5-e8b3-43d7-8525-f42c27e41935)

**Comment by DHeroPlasma** [comment]
Hello, thanks for pointing this out, it fixed the problem with the normal map for some materials in the build.
However, there still seems to be a problem with glass-like materials:

![RTSL_Glass-in-build](https://github.com/Battlehub0x/RTE_Docs/assets/60179146/1c52a1af-5c49-49f7-a493-e968d67f1155)

This might have to do with one of the following properties not being saved correctly:

![Glass_Material-Settings](https://github.com/Battlehub0x/RTE_Docs/assets/60179146/f9b4f5a6-097e-402d-91c8-37f23cc85caa)

Is this something that can be fixed on RTSL-side? Or is this problem on my side?
Could it be that something has to be toggled here in the automatically generated part of the partial class?

![image](https://github.com/Battlehub0x/RTE_Docs/assets/60179146/6f3753d2-0358-4bc2-872b-534c1f0c1d9b)

If you want to try it with the model I am using, you can get it here: https://sketchfab.com/3d-models/toyota-supra-a80-1993-dd897d7823784bc5893c183c1328e8cb

Thank you for your help!

**Comment by Battlehub0x** [comment]
@DHeroPlasma This doesn't seem to be reproducible in the latest RTE 3.5.3.

I tried changing the Universal Render Pipeline\Lit material surface type to Transparent and BaseMap Alpha to 0.5 but it seems to be saved and loaded correctly.

Could you check this, maybe I have some wrong material settings.

![image](https://github.com/Battlehub0x/RTE_Docs/assets/15802443/b65e0cd0-3d2d-47a8-8810-a020af73dc4e)

**Comment by DHeroPlasma** [comment]
Thanks for your answer, I am using an older version of RTE (2.11), but I am currently in the process of upgrading to 3.5.3. Unless you don't have a quick solution to fix this in 2.11, I will leave it as it is.

**Comment by Battlehub0x** [comment]
@DHeroPlasma perhaps a quick fix would be to use URPLitMaterialUtils.SetMaterialKeywords(o)  at the end of the PersistentMaterial.WriteTo method.

![image](https://github.com/Battlehub0x/RTE_Docs/assets/15802443/2e4cad3e-f875-4aef-9b30-6dee080420a3)

```
using Battlehub.RTCommon;
using System;
using UnityEngine;
using UnityEngine.Rendering;

namespace Battlehub.Utils.URP
{
    public enum WorkflowMode
    {
        Specular = 0,
        Metallic
    }

    public enum SmoothnessMapChannel
    {
        SpecularMetallicAlpha,
        AlbedoAlpha,
    }

    public enum SurfaceType
    {
        Opaque = 0,
        Transparent = 1
    }
    public enum BlendMode
    {
        Alpha = 0,
        Premultiply = 1,
        Additive = 2,
        Multiply = 3
    }
    public enum SmoothnessSource
    {
        BaseAlpha = 0,
        SpecularAlpha = 1
    }
    public enum RenderFace
    {
        Both = 0,
        Back = 1,
        Front = 2
    }

    public class URPLitMaterialUtils : IMaterialUtil
    {
        void IMaterialUtil.SetMaterialKeywords(Material material)
        {
            if (material == null || material.shader == null || material.shader.name != RenderPipelineInfo.DefaultShaderName)
            {
                return;
            }

            SetMaterialKeywords(material, SetMaterialKeywords);
            if(material.Color().a == 1)
            {
                material.SetShaderPassEnabled("ShadowCaster", true);
            }
        }
    
        private const int queueOffsetRange = 50;

        public static SmoothnessMapChannel GetSmoothnessMapChannel(Material material)
        {
            int ch = (int)material.GetFloat("_SmoothnessTextureChannel");
            if (ch == (int)SmoothnessMapChannel.AlbedoAlpha)
                return SmoothnessMapChannel.AlbedoAlpha;

            return SmoothnessMapChannel.SpecularMetallicAlpha;
        }

        public static void SetMaterialKeywords(Material material)
        {
            // Note: keywords must be based on Material value not on MaterialProperty due to multi-edit & material animation
            // (MaterialProperty value might come from renderer material property block)
            var hasGlossMap = false;
            var isSpecularWorkFlow = false;
            var opaque = ((SurfaceType)material.GetFloat("_Surface") ==
                          SurfaceType.Opaque);
            if (material.HasProperty("_WorkflowMode"))
            {
                isSpecularWorkFlow = (WorkflowMode)material.GetFloat("_WorkflowMode") == WorkflowMode.Specular;
                if (isSpecularWorkFlow)
                    hasGlossMap = material.GetTexture("_SpecGlossMap") != null;
                else
                    hasGlossMap = material.GetTexture("_MetallicGlossMap") != null;
            }
            else
            {
                hasGlossMap = material.GetTexture("_MetallicGlossMap") != null;
            }

            CoreUtils.SetKeyword(material, "_SPECULAR_SETUP", isSpecularWorkFlow);

            CoreUtils.SetKeyword(material, "_METALLICSPECGLOSSMAP", hasGlossMap);

            if (material.HasProperty("_SpecularHighlights"))
                CoreUtils.SetKeyword(material, "_SPECULARHIGHLIGHTS_OFF",
                    material.GetFloat("_SpecularHighlights") == 0.0f);
            if (material.HasProperty("_EnvironmentReflections"))
                CoreUtils.SetKeyword(material, "_ENVIRONMENTREFLECTIONS_OFF",
                    material.GetFloat("_EnvironmentReflections") == 0.0f);
            if (material.HasProperty("_OcclusionMap"))
                CoreUtils.SetKeyword(material, "_OCCLUSIONMAP", material.GetTexture("_OcclusionMap"));

            if (material.HasProperty("_SmoothnessTextureChannel"))
            {
                CoreUtils.SetKeyword(material, "_SMOOTHNESS_TEXTURE_ALBEDO_CHANNEL_A",
                    GetSmoothnessMapChannel(material) == SmoothnessMapChannel.AlbedoAlpha && opaque);
            }
        }

        public static void SetMaterialKeywords(Material material, Action<Material> shadingModelFunc = null, Action<Material> shaderFunc = null)
        {
            // Clear all keywords for fresh start
            material.shaderKeywords = null;
            // Setup blending - consistent across all Universal RP shaders
            SetupMaterialBlendMode(material);
            // Receive Shadows
            if (material.HasProperty("_ReceiveShadows"))
                CoreUtils.SetKeyword(material, "_RECEIVE_SHADOWS_OFF", material.GetFloat("_ReceiveShadows") == 0.0f);
            // Emission
            //if (material.HasProperty("_EmissionColor"))
                //MaterialEditor.FixupEmissiveFlag(material);
            bool shouldEmissionBeEnabled =
                (material.globalIlluminationFlags & MaterialGlobalIlluminationFlags.EmissiveIsBlack) == 0;
            if (material.HasProperty("_EmissionEnabled") && !shouldEmissionBeEnabled)
                shouldEmissionBeEnabled = material.GetFloat("_EmissionEnabled") >= 0.5f;
            CoreUtils.SetKeyword(material, "_EMISSION", shouldEmissionBeEnabled);
            // Normal Map
            if (material.HasProperty("_BumpMap"))
                CoreUtils.SetKeyword(material, "_NORMALMAP", material.GetTexture("_BumpMap"));
            // Shader specific keyword functions
            shadingModelFunc?.Invoke(material);
            shaderFunc?.Invoke(material);
        }

        public static void SetupMaterialBlendMode(Material material)
        {
            if (material == null)
                throw new ArgumentNullException("material");

            bool alphaClip = material.GetFloat("_AlphaClip") == 1;
            if (alphaClip)
            {
                material.EnableKeyword("_ALPHATEST_ON");
            }
            else
            {
                material.DisableKeyword("_ALPHATEST_ON");
            }

            var queueOffset = 0; // queueOffsetRange;
            if (material.HasProperty("_QueueOffset"))
                queueOffset = queueOffsetRange - (int)material.GetFloat("_QueueOffset");

            SurfaceType surfaceType = (SurfaceType)material.GetFloat("_Surface");
            if (surfaceType == SurfaceType.Opaque)
            {
                if (alphaClip)
                {
                    material.renderQueue = (int)UnityEngine.Rendering.RenderQueue.AlphaTest;
                    material.SetOverrideTag("RenderType", "TransparentCutout");
                }
                else
                {
                    material.renderQueue = (int)UnityEngine.Rendering.RenderQueue.Geometry;
                    material.SetOverrideTag("RenderType", "Opaque");
                }
                material.renderQueue += queueOffset;
                material.SetInt("_SrcBlend", (int)UnityEngine.Rendering.BlendMode.One);
                material.SetInt("_DstBlend", (int)UnityEngine.Rendering.BlendMode.Zero);
                material.SetInt("_ZWrite", 1);
                material.DisableKeyword("_ALPHAPREMULTIPLY_ON");
                material.SetShaderPassEnabled("ShadowCaster", true);
            }
            else
            {
                BlendMode blendMode = (BlendMode)material.GetFloat("_Blend");
                var queue = (int)UnityEngine.Rendering.RenderQueue.Transparent;

                // Specific Transparent Mode Settings
                switch (blendMode)
                {
                    case BlendMode.Alpha:
                        material.SetInt("_SrcBlend", (int)UnityEngine.Rendering.BlendMode.SrcAlpha);
                        material.SetInt("_DstBlend", (int)UnityEngine.Rendering.BlendMode.OneMinusSrcAlpha);
                        material.DisableKeyword("_ALPHAPREMULTIPLY_ON");
                        break;
                    case BlendMode.Premultiply:
                        material.SetInt("_SrcBlend", (int)UnityEngine.Rendering.BlendMode.One);
                        material.SetInt("_DstBlend", (int)UnityEngine.Rendering.BlendMode.OneMinusSrcAlpha);
                        material.EnableKeyword("_ALPHAPREMULTIPLY_ON");
                        break;
                    case BlendMode.Additive:
                        material.SetInt("_SrcBlend", (int)UnityEngine.Rendering.BlendMode.SrcAlpha);
                        material.SetInt("_DstBlend", (int)UnityEngine.Rendering.BlendMode.One);
                        material.DisableKeyword("_ALPHAPREMULTIPLY_ON");
                        break;
                    case BlendMode.Multiply:
                        material.SetInt("_SrcBlend", (int)UnityEngine.Rendering.BlendMode.DstColor);
                        material.SetInt("_DstBlend", (int)UnityEngine.Rendering.BlendMode.Zero);
                        material.DisableKeyword("_ALPHAPREMULTIPLY_ON");
                        material.EnableKeyword("_ALPHAMODULATE_ON");
                        break;
                }
                // General Transparent Material Settings
                material.SetOverrideTag("RenderType", "Transparent");
                material.SetInt("_ZWrite", 0);
                material.renderQueue = queue + queueOffset;
                material.SetShaderPassEnabled("ShadowCaster", false);
            }
        }
    }
}
```
